src/emit/substrate/target.rs

```diff
@@ -1580,14 +1580,73 @@ impl<'a> TargetRuntime<'a> for SubstrateTarget {
 
     fn builtin_function(
         &self,
-        _binary: &Binary<'a>,
-        _function: FunctionValue<'a>,
-        _builtin_func: &Function,
-        _args: &[BasicMetadataValueEnum<'a>],
+        binary: &Binary<'a>,
+        function: FunctionValue<'a>,
+        builtin_func: &Function,
+        args: &[BasicMetadataValueEnum<'a>],
         _first_arg_type: BasicTypeEnum,
         _ns: &Namespace,
     ) -> BasicValueEnum<'a> {
-        unimplemented!()
+        emit_context!(binary);
+
+        assert_eq!(builtin_func.name, "chain_extension", "unimplemented");
+
+        let input_ptr = binary.vector_bytes(args[1].into_pointer_value().into());
+        let input_len = binary.vector_len(args[1].into_pointer_value().into());
+        let (output_ptr, output_len_ptr) = scratch_buf!();
+        let len = 16384;
+        binary.builder.build_store(output_len_ptr, i32_const!(len));
+        call!(
+            "__memset8",
+            &[
+                output_ptr.into(),
+                binary.context.i64_type().const_zero().into(),
+                i32_const!(len / 8).into()
+            ]
+        );
+        let ret = call!(
+            "call_chain_extension",
+            &[
+                args[0].into_int_value().into(),
+                input_ptr.into(),
+                input_len.into(),
+                output_ptr.into(),
+                output_len_ptr.into()
+            ]
+        )
+        .try_as_basic_value()
+        .left()
+        .unwrap()
+        .into_int_value();
+
+        log_return_code(binary, "call_chain_extension", ret);
+
+        let ok = binary.context.append_basic_block(function, "ok_block");
+        let done = binary.context.append_basic_block(function, "done_block");
+        let success = binary
+            .builder
+            .build_int_compare(IntPredicate::EQ, ret, i32_zero!(), "ret");
+        binary.builder.build_conditional_branch(success, ok, done);
```

Comment:
What is supposed to happen if the return value from the syscall is not 0?

Reply:
[It'll return it to the caller](https://github.com/hyperledger/solang/blob/f85e43ba643e67826dffca0910202b6134db29fd/src/emit/instructions.rs#LL548C50-L548C50)

However now that I think of it, it might be better to just trap. One could come up with the idea to `try-catch` it, but according to the solidity docs this is only supposed to be used for external calls

Reply:
Actually I don't know about that logic in `instruction.rs`. Is it guaranteed that we always stay in `call` or `deploy` and this return there will return from the contract? I'm not sure about this. I think we should just trap.

Reply:
I changed the behavior entirely. We should not trap or do anything on a non-zero return value. Instead, we should return this value back to the caller.

In the case of `call_chain_extension`, the return value being non-zero has no apparent meaning. It depends on the chain extensions concrete implementation. Hence we should just return it to the caller, as he should know how to handle this.